Feature/wysiwyg editor

Arkitecht opened the pull request:
This adds the ability to enable rich text formatting using a WYSIWYG editor.

tomirons commented on resources/views/components/post-item.blade.php:
```diff
@@ -34,6 +34,10 @@
     </div>
 
     <div class="mt-2">
+        @if ($post->formatting == 'rich')
+        {!! $post->content !!}
+        @else
         {!! nl2br(e($post->content)) !!}
+        @endif
```
Would this work?
```suggestion
        {!! $post->formatting == 'rich' ? $post->content : nl2br(e($post->content)) !!}
```

tomirons replied:
Also curious if the `$post->formatting == 'rich` could be put into a method on the model itself? Or even this entire thing as an accessor?

Arkitecht replied:
Good call, I have added this as an attribute on the Post model, and simplified the logic with the ternary operator

tomirons commented on resources/views/layouts/app.blade.php:
```diff
@@ -14,6 +14,10 @@
     <script src="{{ asset('vendor/firefly/js/app.js') }}" defer></script>
+    @if (config('firefly.features.wysiwyg'))
+    <script src="//cdn.quilljs.com/1.3.6/quill.min.js"></script>
+    <link href="https://cdn.quilljs.com/1.3.6/quill.snow.css" rel="stylesheet">
```
Should we allow the theme to be configurable if quill offers more than just _snow_?

Arkitecht replied:
I was thinking about that. There are a lot of configurability options, in terms of theming, and also the toolbar / buttons that are visible. My thought process was that this is what is provided "out of the box" and there is always the option to publish and then override the views, if one wants to customize the theme or buttons.


Another idea would be to make features => wysiwyg an array, rather than a boolean, something like this:
```php
'features' => [
        'watchers' => true,
        'wysiwyg'  => [
            'theme' => 'snow',
            'toolbar_options' => []
        ],
        'correct_posts' => true
    ],
```
 Thoughts?

AndreasElia replied:
Sorry, I didn't see this, but yes, I agree with putting in a config like that.

tomirons replied:
I would also include an `enabled` key in that array.

Arkitecht replied:
I have implemented this

tomirons commented on resources/views/components/quill-js.blade.php:
```diff
@@ -0,0 +1,35 @@
+        };
+        var editor = new Quill('#content',options);
+
+        document.getElementById('submit').addEventListener('click',function(event) {
+            var content = document.getElementsByClassName('ql-editor')[0].innerHTML
+            document.getElementById('content_hidden').setAttribute('value', content)
+        })
```
What is this for exactly? 🤔

Arkitecht replied:
Weirdly enough, Quill provides the WYSIWYG content editor, but uses a div, not a textarea, so does not hook into the form submission directly. This takes the content from Quill, and puts it in a hidden form field on submit to send it back o the server with the form submission.

tomirons reviewed (changes requested):
This is a very nice feature! 👍  I just have a few questions/concerns.

AndreasElia commented on resources/views/components/quill-js.blade.php:
```diff
@@ -0,0 +1,35 @@
+        };
+        var editor = new Quill('#content',options);
+
+        document.getElementById('submit').addEventListener('click',function(event) {
```
Lacks spacing here `'click',function`

AndreasElia commented on resources/views/components/quill-js.blade.php:
```diff
@@ -0,0 +1,35 @@
+            },
+            theme: 'snow'
+        };
+        var editor = new Quill('#content',options);
```
Lacks spacing here `'#content',options`

AndreasElia commented on resources/views/components/rich-textarea.blade.php:
```diff
@@ -0,0 +1,9 @@
+    <input id="formatting" type="hidden" name="formatting" value="rich">
+    <div id="content">{!! old('content', $value) !!}</div>
+@else
+    <x-textarea id="content" class="block mt-1 w-full" type="text" name="content" required autofocus>{{ old('content', $value) }}</x-textarea>
```
Line too long, break this onto 3 lines please.

```blade
<x-textarea id="content" class="block mt-1 w-full" type="text" name="content" required autofocus>
    {{ old('content', $value) }}
</x-textarea>
```

AndreasElia commented:
Nice work 😄 Could you post a screenshot into this PR of what this looks like too, please? Both the WYSIWYG editor, and how a formatted post can look.

Arkitecht commented:
@AndreasElia - I updated the code formatting per your requests. In preparing the screenshot I also realize Tailwind "resets" most built in formatting, so I added in an unreset for the rich text view to allow that back in.

Screenshot of formatting and editor in action:
![image](https://user-images.githubusercontent.com/66117/124975688-42c6a780-dffc-11eb-8c58-d4a2fb61a584.png)

AndreasElia commented:
Good stuff 👍  Do you think `toolbarOptions` should be in the config file, and kept minimal by default (e.g. only bold, underline, italic and lists by default)?

Arkitecht commented:
Please see my comment and question above (https://github.com/AndreasElia/laravel-firefly/pull/46#discussion_r666394393) regarding the toolbar / config. I am on-board with those as the defaults

AndreasElia commented:
Replied to that thread, but just to repeat that the default toolbar options should be minimal as it is a forum rather than a CMS.

tomirons commented on resources/views/components/post-item.blade.php:
```diff
@@ -34,10 +34,6 @@
-        @else
-        {!! nl2br(e($post->content)) !!}
-        @endif
+        {!! $post->isRichlyFormatted ? $post->content : nl2br(e($post->content)) !!}
```
What about doing `getContentAttribute()` and adding this to it?

Arkitecht replied:
I don't love overriding an actual column name with an attribute value, how do you feel about `getFormattedContentAttribute()`?

Arkitecht commented:
Here is the updated UI with the simpler toolbar (and 1 added custom button, to add font color)

![Screen Shot 2021-07-09 at 9 24 16 AM](https://user-images.githubusercontent.com/66117/125085856-06955480-e099-11eb-8082-193beb2d0283.png)

Arkitecht commented:
When reviewing with my team one of my teammates called out that there should be some XSS protection added in. 

Added a commit which strips script tags from the submission. quill seems to handle this from the front-end OK, but a clever hacker could bypass it, and submit to the back-end

AndreasElia commented on config/firefly.php:
```diff
@@ -93,5 +93,15 @@
         'watchers' => false,
+        'wysiwyg' => [
+            'enabled' => false,
+            'theme' => 'snow', //More about themes at https://quilljs.com/docs/themes/
```
Spacing after `//`

AndreasElia commented on config/firefly.php:
```diff
@@ -93,5 +93,15 @@
+            'toolbar_options' => [ //Docs at https://quilljs.com/docs/modules/toolbar/
+                ['bold', 'italic', 'underline', 'strike'],
+                [['list' => 'ordered'], ['list'=> 'bullet']],
+
```
Remove blank line, options should be okay together, I think?

AndreasElia commented on resources/views/components/quill-js.blade.php:
```diff
@@ -0,0 +1,17 @@
+        };
+        var editor = new Quill('#content', options);
+
+        document.getElementById('submit').addEventListener('click', function(event) {
```
`function (event) {`

AndreasElia commented on resources/views/components/quill-js.blade.php:
```diff
@@ -0,0 +1,17 @@
+            modules: {
+                toolbar: toolbarOptions
+            },
+            theme: '{{config('firefly.features.wysiwyg.theme')}}'
```
Spacing -> `{{ config('firefly.features.wysiwyg.theme') }}`

AndreasElia commented on resources/views/layouts/app.blade.php:
```diff
@@ -14,6 +14,10 @@
     <script src="{{ asset('vendor/firefly/js/app.js') }}" defer></script>
+    @if (config('firefly.features.wysiwyg.enabled'))
+    <script src="//cdn.quilljs.com/1.3.6/quill.min.js"></script>
+    <link href="https://cdn.quilljs.com/1.3.6/quill.{{config('firefly.features.wysiwyg.theme')}}.css" rel="stylesheet">
```
Spacing -> `{{ config('firefly.features.wysiwyg.theme') }}`

AndreasElia commented:
> When reviewing with my team one of my teammates called out that there should be some XSS protection added in.
> 
> Added a commit which strips script tags from the submission. quill seems to handle this from the front-end OK, but a clever hacker could bypass it, and submit to the back-end

Nice work @Arkitecht 👍
